src/Collections/Sequences/Actions.dfy

```diff
@@ -0,0 +1,261 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+/*******************************************************************************
+*  Original: Copyright (c) 2020 Secure Foundations Lab
+*  SPDX-License-Identifier: MIT
+*  
+*  Modifications and Extensions: Copyright by the contributors to the Dafny Project
+*  SPDX-License-Identifier: MIT 
+*******************************************************************************/
+
+include "../../Wrappers.dfy"
+include "../../Action.dfy"
+include "Seq.dfy"
+
+module Actions {
+  import opened Wrappers
+  import opened A = Action
+  import opened Seq
+
+  method Map<A, R>(
+    action: Action<A, R>,
+    s: seq<A>
+  )
+    returns (res: seq<R>)
+    requires forall i | i in s :: action.Requires(i)
+    ensures |s| == |res|
+    ensures
+      forall i ::
+        && 0 <= i < |s|
+      ==>
+        action.Ensures(s[i], res[i])
+  {
+    var rs := [];
+    for i := 0 to |s|
+      invariant |s[..i]| == |rs|
+      invariant forall j ::
+        && 0 <= j < i
+      ==>
+        action.Ensures(s[j], rs[j])
+    {
+      var r := action.Invoke(s[i]);
+      rs := rs + [r];
+    }
+    return rs;
+  }
+
+  method MapWithResult<A, R, E>(
+    action: Action<A, Result<R, E>>,
+    s: seq<A>
+  )
+    returns (res: Result<seq<R>, E>)
+    requires forall i | i in s :: action.Requires(i)
+    ensures
+      res.Success?
+    ==>
+      |s| == |res.value|
+    ensures
+      res.Success?
+    ==>
+        (forall i ::
+          && 0 <= i < |s|
+        ==>
+          action.Ensures(s[i], Success(res.value[i])))
+  {
+    var rs := [];
+    for i := 0 to |s|
+      invariant |s[..i]| == |rs|
+      invariant forall j ::
+        && 0 <= j < i
+      ==>
+        action.Ensures(s[j], Success(rs[j]))
+    {
+      var maybeR := action.Invoke(s[i]);
+      if maybeR.Failure? {
+        return Failure(maybeR.error);
+      }
```

Comment:
Worth a comment about how you're having to work around the issue with `:-` here?